pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp

```diff
@@ -632,5 +636,128 @@ class StateVectorLQubit : public StateVectorBase<PrecisionT, Derived> {
 
         applyMatrix(matrix.data(), wires, inverse);
     }
+
+    /**
+     * @brief Apply a mid-circuit measurement.
+     *
+     * @param wires Wires to sample from.
+     * @param postselect Value to postselect (0 or 1) or empty to ignore
+     * @param reset Flag to reset wire to 0
+     */
+    inline auto
+    applyMidMeasureMP(const std::vector<std::size_t> &wires,
+                      const std::vector<std::size_t> &postselect = {},
+                      bool reset = false) -> int {
+        PL_ABORT_IF_NOT(wires.size() == 1,
+                        "MidMeasureMP should have a single wire.")
+        PL_ABORT_IF(postselect.size() > 1,
+                    "MidMeasureMP accepts at most one postselect value.")
+        const int ps = postselect.empty() ? -1 : postselect[0];
+
+        std::vector<PrecisionT> probs_ = probs(wires[0]);
+        auto sample = randomSample(probs_[0]);
+        if (ps != -1 && ps != sample) {
+            sample = -1;
```

Comment:
Does this mean the operation can return a measurement result of -1?

Reply:
How do you deal with filtering out a shot in Catalyst?

Reply:
The short answer is: we don't 😅 

QNode execution in Catalyst is currently exclusively managed by the user program (user program here means "all binary code staged out from python"). The Catalyst runtime environment merely facilitates communication between the user program and the quantum device, but does not manage execution.

I've provided some more context offline so we might need to come a decision on that discussion first before determining whether the proposed implementation here is compatible with Catalyst (at the moment it isn't).

Reply:
Yes, it is in the Python layer [here](https://github.com/PennyLaneAI/pennylane-lightning/pull/621/files/7fb6faaba1becce5a38cdd244f745d2e575536d6#diff-068e2aa5237a375b6af67a124f81018aa46c2d335fecf3035346f0627e6ef7a4R454) and then in PennyLane.

Reply:
Well if this is used in PennyLane the incompatibility goes a little bit deeper. We need to think about that.